Virt handler fixes and cleanup

fossedihelm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
The virt-handler suffers of multiple possible race conditions during shutdown:
1. There isn't a correct sync between the controller and the heartbeat shutdown
2. There isn't proper error handling of the final node labelling process.

#### For the first one:
The intent of the `heartBeatDone` channel was
to ensure the heartbeat is finished
before stopping the handler vm controller.
In particular, its intention was to prevent
a shutdown race:
- heartbeat tries to stop and do a last http request
- controller shuts down
If the controller shuts down first, then the main function
just exists, killing all remaining go routines right away.

Actually it is not doing its job since
`c.heartBeat.Run(c.heartBeatInterval, stopCh)` is
not a blocking function (it spawns a goroutine and
immediately returns the channel).
As result, the `heartBeat` is always immediately
closed after the start of the heartBeat goroutine.

To ensure that the heartbeat is completed before shutting
down the controller, we should block on the channel returned
by the heartbeat itself.


#### For the second one:
When the heartbeat is notified of a shutdown
by the stopchan, it marks the node unschedulable
before returning and thus end the entire process.

This labelling process is performed in a goroutine
which at the end signal the job done closing the
`done` channel.

This logic actually is not necessary, since it can
be a simple body function, without the goroutine.

In addition, there isn't a proper error handling.
In an error occurs during the goroutine execution
(i.e. node patch failure), nothing is returned
and the channel is never closed, which cause the
`<-labelNodeDone` to hangs and leave the virt-handler
stuck.

With this commit both problem are handler:
1. drop the unneeded goroutine
2. handle the error retrying it for a `DefaultBackoff`
(https://github.com/kubernetes/client-go/blob/37045084c2aa82927b0e5ffc752861430fd7e4ab/util/retry/util.go#L38)


Lastly, there is an optimization on the guest agent command support
calculation.

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
4. If no release note is required, just write "NONE".
-->
```release-note
Bugfix: fix possible virt-handler race condition and stuck situation during shutdown
```

fossedihelm commented:
/cc @jean-edouard @vladikr @iholder101

fossedihelm commented:
/uncc @lyarwood  @victortoso

fossedihelm commented:
/test pull-kubevirt-e2e-k8s-1.31-sig-compute-serial

iholder101 commented:
/kind cleanup

fossedihelm commented on pkg/virt-handler/vm.go:
```diff
@@ -1712,14 +1701,14 @@ func (cond *multipleNonMigratableCondition) generateStorageLiveMigrationConditio
 }
 
-func (d *VirtualMachineController) calculateLiveStorageMigrationCondition(vmi *v1.VirtualMachineInstance) *v1.VirtualMachineInstanceCondition {
+func (c *VirtualMachineController) calculateLiveStorageMigrationCondition(vmi *v1.VirtualMachineInstance) *v1.VirtualMachineInstanceCondition {
```
Yes, :slightly_smiling_face: just `c` for controller

fossedihelm commented on pkg/virt-handler/guest-agent/support.go:
```diff
@@ -53,28 +53,22 @@ var PasswordRelatedGuestAgentCommands = []string{
-			}
-		}
-		if found == false {
+		if enabled, exists := availableCmdsMap[cmd]; !exists || !enabled {
```
Well, we should return true if every required command is also available and enabled.
Currently, the `if` is looking for something that is not available or enabled and early exit as soon as finds a match.
Your suggestion will return true as soon as one (positive) match is found, which does not guarantee that all the
commands are available and enabled  :)

iholder101 replied:
Ah :man_facepalming: yeah I missed that. You're right, thanks!

fossedihelm commented on pkg/virt-handler/guest-agent/support.go:
I am not sure. Actually they are already tested in `vm_test`

fossedihelm commented on pkg/virt-handler/vm.go:
```diff
@@ -1677,11 +1677,7 @@ func (c *VirtualMachineController) Run(threadiness int, stopCh chan struct{}) {
-		c.heartBeat.Run(c.heartBeatInterval, stopCh)
-		close(heartBeatDone)
-	}()
+	heartBeatDone := c.heartBeat.Run(c.heartBeatInterval, stopCh)
```
I think we need it to guarantee that the `h.labelNodeUnschedulable()` is executed when the `stop` channel is closed.
The VM controller runs the heartBeat through a goroutine and continues its execution.

**-- Without the `heartBeatDone` --**
When the stop channel is closed, both heartBeat goroutine and VM controller continue their execution in parallel:
- the controller tries to end the whole function
- the heartBeat marks the node unschedulable and ends.

This is a race condition since there are two things that could happen:

- the heartBeat is fast enough and ends before the controller ends
   - node is properly marked unschedulable
  
- the controller ends before the heartBeat finishes its job and gets killed since the parent process is dead.
   - the heartBeat may or may have not marked the node unschedulable
   
**-- With the `heartBeatDone` --**
When the stop channel is closed, the VM controller waits for the end of the heartBeat before continue its execution and shutdown.

fossedihelm replied:
Does it make sense?

iholder101 replied:
Yes, it does :) thank you!

fossedihelm commented:
[Addressed comments](https://github.com/kubevirt/kubevirt/compare/2f535214199b5c9d5ae99ae0a1009e9d07e2897a..8d3a14dbf4ad86e39261c96955ff050ce5a80f72) and [Rebase](https://github.com/kubevirt/kubevirt/compare/8d3a14dbf4ad86e39261c96955ff050ce5a80f72..06eae13c47dab8f1c90b8385bfe3b60a3f1ad3c2)
@iholder101 PTAL, thanks

jean-edouard reviewed (commented):
/approve
Very nice, thanks!
Just one question, why `guest-agent/support.go` and not just `guest-agent.go`? Do we plan on adding more files to that directory?
Also if the guest agent functions were in the same package as the rest of virt-handler, we wouldn't need to prefix all calls with `guest_agent.` (is it right to use an underscore there btw? Looks like an unusual package name)

fossedihelm commented:
> /approve Very nice, thanks! Just one question, why `guest-agent/support.go` and not just `guest-agent.go`? Do we plan on adding more files to that directory? Also if the guest agent functions were in the same package as the rest of virt-handler, we wouldn't need to prefix all calls with `guest_agent.` (is it right to use an underscore there btw? Looks like an unusual package name)

@jean-edouard Right! I will drop the directory in favor of a simple file :+1: Thank you!

iholder101 commented on pkg/virt-handler/vm.go:
```diff
@@ -1704,8 +1700,8 @@ func (c *VirtualMachineController) Run(threadiness int, stopCh chan struct{}) {
 	}
 
-	<-stopCh
 	<-heartBeatDone
+	<-stopCh
```
nit: why is this needed? I'm pretty sure that the behavior is equivalent to before your changes.

fossedihelm replied:
Yes, the behavior is equivalent.
I do that to highlight that the heartbeat should succeed before moving on. :)
Thank you!
I will rollback if you think it's misleading

iholder101 reviewed (commented):
Thank you @fossedihelm!
/lgtm

brianmcarey commented:
/override pull-kubevirt-e2e-k8s-1.31-sig-performance

Triggered by status-reconciler when lane was bumped.
